File: packages/protocol-ts/tests/validate.test.ts
/**
 * Tests del validador `validateComposition` — KRO-79.
 *
 * Cubre las reglas principales con casos happy + fail. La estructura es:
 *
 *   - "happy path" — composition canónica → valid=true, sin issues
 *   - Por cada regla, un caso fail aislado → issue específica
 *   - Tests semánticos: warnings no invalidan, sin fieldDefs no falla
 *     por referencias
 */

import { describe, it, expect } from 'vitest';
import { validateComposition } from '../src/validate';
import type { ViewComposition, FieldDefLike } from '../src/index';

// ── Helpers ────────────────────────────────────────────────────────

/** Composition canónica simple: hero_protagonico con title=nombre. */
function makeValidComposition(): ViewComposition {
  return {
    recipe: 'hero_protagonico',
    action: 'none',
    slots: {
      title: { fields: ['nombre'] },
    },
  };
}

function fieldsBase(): FieldDefLike[] {
  return [
    { key: 'nombre',  type: 'text' },
    { key: 'edad',    type: 'number' },
    { key: 'avatar',  type: 'image' },
    { key: 'galeria', type: 'array<image>', behavior: 'gallery' },
  ];
}

// ── Happy path ─────────────────────────────────────────────────────

describe('validateComposition — happy paths', () => {
  it('composition canónica sin fieldDefs → valid', () => {
    const r = validateComposition(makeValidComposition());
    expect(r.valid).toBe(true);
    expect(r.issues).toEqual([]);
  });

  it('composition canónica con fieldDefs → valid', () => {
    const r = validateComposition(makeValidComposition(), { fieldDefs: fieldsBase() });
    expect(r.valid).toBe(true);
    expect(r.issues.filter(i => i.level === 'error')).toEqual([]);
  });

  it('composition con accentPosition válido → valid', () => {
    const c = makeValidComposition();
    c.accentPosition = 'top';
    const r = validateComposition(c);
    expect(r.valid).toBe(true);
  });
});

// ── recipe ─────────────────────────────────────────────────────────

describe('validateComposition — recipe', () => {
  it('recipe inexistente → error', () => {
    const c = makeValidComposition();
    (c as any).recipe = 'fake_recipe';
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues).toContainEqual(expect.objectContaining({
      path: 'recipe',
      level: 'error',
    }));
  });
});

// ── action ─────────────────────────────────────────────────────────

describe('validateComposition — action', () => {
  it('action inválida → error', () => {
    const c = makeValidComposition();
    (c as any).action = 'fake_action';
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues).toContainEqual(expect.objectContaining({
      path: 'action',
      level: 'error',
    }));
  });

  it('cada action válida pasa', () => {
    const actions = ['none', 'navigate_to_detail', 'modal', 'expand_inline', 'external_link'] as const;
    actions.forEach(a => {
      const c = makeValidComposition();
      c.action = a;
      const r = validateComposition(c);
      const errs = r.issues.filter(i => i.path === 'action' && i.level === 'error');
      expect(errs).toEqual([]);
    });
  });
});

// ── slot fields ────────────────────────────────────────────────────

describe('validateComposition — slot fields', () => {
  it('field inexistente con fieldDefs → error', () => {
    const c = makeValidComposition();
    c.slots.title.fields = ['fantasma'];
    const r = validateComposition(c, { fieldDefs: fieldsBase() });
    expect(r.valid).toBe(false);
    expect(r.issues.some(i => i.message.includes('fantasma'))).toBe(true);
  });

  it('field inexistente SIN fieldDefs → no falla (no se valida)', () => {
    const c = makeValidComposition();
    c.slots.title.fields = ['fantasma'];
    const r = validateComposition(c);
    expect(r.issues.filter(i => i.path.startsWith('slots.title.fields'))).toEqual([]);
  });

  it('field incompatible con accept kind → error', () => {
    // title acepta text-short. Pasamos un field type=image → incompatible.
    const c = makeValidComposition();
    c.slots.title.fields = ['avatar'];
    const r = validateComposition(c, { fieldDefs: fieldsBase() });
    expect(r.valid).toBe(false);
    expect(r.issues.some(i => i.message.includes('no es compatible'))).toBe(true);
  });

  it('single slot con múltiples fields → warn (no error)', () => {
    const c = makeValidComposition();
    c.slots.title.fields = ['nombre', 'edad']; // title es single
    const r = validateComposition(c, { fieldDefs: fieldsBase() });
    const warns = r.issues.filter(i => i.level === 'warn');
    expect(warns.some(w => w.message.includes('single'))).toBe(true);
  });
});

// ── appearance ─────────────────────────────────────────────────────

describe('validateComposition — appearance', () => {
  it('shape inválido → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { shape: 'fake' as any };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues).toContainEqual(expect.objectContaining({
      path: 'slots.title.appearance.shape',
      level: 'error',
    }));
  });

  it('aspect inválido → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { aspect: 'fake' as any };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('align inválido → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { align: 'fake' as any };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('weight inválido → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { weight: 'fake' as any };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('size inválido → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { size: 'fake' as any };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('truncateChars fuera de rango → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { truncateChars: 1000 };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);

    const c2 = makeValidComposition();
    c2.slots.title.appearance = { truncateChars: 0 };
    const r2 = validateComposition(c2);
    expect(r2.valid).toBe(false);
  });

  it('truncateChars válido (1-500) → ok', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { truncateChars: 80 };
    const r = validateComposition(c);
    expect(r.valid).toBe(true);
  });

  it('accentPosition inválido → error', () => {
    const c = makeValidComposition();
    c.slots.title.appearance = { accentPosition: 'fake' as any };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('accentPosition top-level inválido → error', () => {
    const c = makeValidComposition();
    (c as any).accentPosition = 'fake';
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues).toContainEqual(expect.objectContaining({
      path: 'accentPosition',
      level: 'error',
    }));
  });
});

// ── orientation ────────────────────────────────────────────────────

describe('validateComposition — orientation', () => {
  it('orientation inválida → error', () => {
    const c = makeValidComposition();
    (c.slots.title as any).orientation = 'diagonal';
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('orientation horizontal/vertical → ok', () => {
    const c = makeValidComposition();
    c.slots.title.orientation = 'vertical';
    const r = validateComposition(c);
    expect(r.valid).toBe(true);
  });
});

// ── expand ─────────────────────────────────────────────────────────

describe('validateComposition — expand', () => {
  it('expand.recipe inexistente → error', () => {
    const c = makeValidComposition();
    c.expand = { recipe: 'fake' as any, slots: {} };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('expand.recipe de kind=detail → error', () => {
    const c = makeValidComposition();
    c.expand = { recipe: 'hero_protagonico', slots: {} }; // hero es detail, no expand
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues.some(i => i.message.includes('no es de kind=expand'))).toBe(true);
  });

  it('expand.recipe de kind=expand → ok', () => {
    const c = makeValidComposition();
    c.expand = { recipe: 'accordion_simple', slots: {} };
    const r = validateComposition(c);
    expect(r.issues.filter(i => i.path.startsWith('expand.recipe') && i.level === 'error')).toEqual([]);
  });
});

// ── targetRecipe ───────────────────────────────────────────────────

describe('validateComposition — targetRecipe', () => {
  it('targetRecipe inexistente → error', () => {
    const c = makeValidComposition();
    (c as any).targetRecipe = 'fake';
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('targetRecipe de kind=list → warn (no error)', () => {
    const c = makeValidComposition();
    c.targetRecipe = 'compact_avatar'; // list, no detail
    const r = validateComposition(c);
    const warns = r.issues.filter(i => i.path === 'targetRecipe' && i.level === 'warn');
    expect(warns.length).toBeGreaterThan(0);
    expect(r.valid).toBe(true);
  });

  it('targetRecipe de kind=detail → ok', () => {
    const c = makeValidComposition();
    c.targetRecipe = 'hero_protagonico';
    const r = validateComposition(c);
    expect(r.issues.filter(i => i.path === 'targetRecipe')).toEqual([]);
  });
});

// ── slotOverrides ──────────────────────────────────────────────────

describe('validateComposition — slotOverrides', () => {
  it('disabled con slot inexistente → warn', () => {
    const c = makeValidComposition();
    c.slotOverrides = { disabled: ['fantasma'] };
    const r = validateComposition(c);
    const warns = r.issues.filter(i => i.level === 'warn');
    expect(warns.some(w => w.path.startsWith('slotOverrides.disabled'))).toBe(true);
    expect(r.valid).toBe(true); // warns no invalidan
  });

  it('custom slot shape válido → ok', () => {
    const c = makeValidComposition();
    c.slotOverrides = {
      custom: [{
        id:      'extra',
        label:   'Extra',
        kind:    'single',
        accepts: ['text-short'],
      }],
    };
    const r = validateComposition(c);
    expect(r.issues.filter(i => i.path.startsWith('slotOverrides.custom'))).toEqual([]);
  });

  it('custom slot sin label → error', () => {
    const c = makeValidComposition();
    c.slotOverrides = {
      custom: [{ id: 'x', label: '', kind: 'single', accepts: ['text-short'] }],
    };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('custom slot con kind inválido → error', () => {
    const c = makeValidComposition();
    c.slotOverrides = {
      custom: [{ id: 'x', label: 'X', kind: 'fake' as any, accepts: ['text-short'] }],
    };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });
});

// ── linkField (KRO-80) ─────────────────────────────────────────────

describe('validateComposition — linkField', () => {
  it('linkField ausente → ok', () => {
    const r = validateComposition(makeValidComposition(), { fieldDefs: fieldsBase() });
    expect(r.issues.filter(i => i.path === 'linkField')).toEqual([]);
  });

  it('linkField string vacío → error', () => {
    const c = makeValidComposition();
    c.linkField = '   ';
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues.some(i => i.path === 'linkField')).toBe(true);
  });

  it('linkField referencia field existente → ok', () => {
    const c = makeValidComposition();
    c.linkField = 'nombre';
    const r = validateComposition(c, { fieldDefs: fieldsBase() });
    expect(r.issues.filter(i => i.path === 'linkField')).toEqual([]);
  });

  it('linkField referencia field inexistente → error', () => {
    const c = makeValidComposition();
    c.linkField = 'fantasma';
    const r = validateComposition(c, { fieldDefs: fieldsBase() });
    expect(r.valid).toBe(false);
    expect(r.issues.some(i =>
      i.path === 'linkField' && i.message.includes('fantasma'),
    )).toBe(true);
  });

  it('linkField inexistente SIN fieldDefs → no falla (no se valida)', () => {
    const c = makeValidComposition();
    c.linkField = 'fantasma';
    const r = validateComposition(c);
    expect(r.issues.filter(i => i.path === 'linkField')).toEqual([]);
  });
});

// ── nestedComposition (KRO-80) ─────────────────────────────────────

describe('validateComposition — nestedComposition (depth max=2)', () => {
  it('nested válido (depth 2) → ok', () => {
    const c = makeValidComposition();
    c.slots.title.nestedComposition = {
      recipe: 'compact_avatar',
      slots: {
        title: { fields: ['otro'] },
      },
    };
    const r = validateComposition(c);
    expect(r.issues.filter(i => i.path.includes('nestedComposition') && i.level === 'error')).toEqual([]);
  });

  it('nested con recipe vacío → error', () => {
    const c = makeValidComposition();
    c.slots.title.nestedComposition = {
      recipe: '' as any,
      slots: {},
    };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues.some(i => i.path.endsWith('nestedComposition.recipe'))).toBe(true);
  });

  it('nested DENTRO de nested → error (depth > 2)', () => {
    const c = makeValidComposition();
    c.slots.title.nestedComposition = {
      recipe: 'compact_avatar',
      slots: {
        title: {
          fields: ['x'],
          nestedComposition: {
            recipe: 'compact_card',
            slots: {},
          },
        },
      },
    };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
    expect(r.issues.some(i => i.message.includes('profundidad máxima'))).toBe(true);
  });

  it('nested.slots con field key vacío → error', () => {
    const c = makeValidComposition();
    c.slots.title.nestedComposition = {
      recipe: 'compact_avatar',
      slots: {
        title: { fields: ['  '] },
      },
    };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('nested.slots NO es objeto → error', () => {
    const c = makeValidComposition();
    c.slots.title.nestedComposition = {
      recipe: 'compact_avatar',
      slots: null as any,
    };
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });
});

// ── Semánticos ─────────────────────────────────────────────────────

describe('validateComposition — semánticos', () => {
  it('valid=true cuando solo hay warnings', () => {
    const c = makeValidComposition();
    c.slotOverrides = { disabled: ['fantasma'] }; // warn
    c.targetRecipe = 'compact_avatar'; // warn (kind wrong)
    const r = validateComposition(c);
    expect(r.issues.length).toBeGreaterThan(0);
    expect(r.issues.every(i => i.level === 'warn')).toBe(true);
    expect(r.valid).toBe(true);
  });

  it('valid=false si hay AL MENOS un error', () => {
    const c = makeValidComposition();
    (c as any).recipe = 'fake'; // 1 error
    c.slotOverrides = { disabled: ['fantasma'] }; // 1 warn
    const r = validateComposition(c);
    expect(r.valid).toBe(false);
  });

  it('cada issue incluye path + message + level', () => {
    const c = makeValidComposition();
    (c as any).recipe = 'fake';
    const r = validateComposition(c);
    r.issues.forEach(i => {
      expect(i.path).toBeTruthy();
      expect(i.message).toBeTruthy();
      expect(['error', 'warn']).toContain(i.level);
    });
  });
});

File: packages/core_dart/test/validate_test.dart
/// Corpus de `validate.dart` — ESPEJO de `tests/validate.test.ts` (KRO-79).
///
/// Construye las composiciones vía constructores (el modelo Dart es tipado).
/// Los casos *estructurales* del TS que en Dart son irrepresentables por el
/// tipo (p.ej. `slots: null`) se omiten con nota: el parse (`fromJson`) los
/// previene antes de llegar al validador — garantía más fuerte, no más débil.
import 'package:kromia_core/kromia_core.dart';
import 'package:test/test.dart';

ViewComposition vc({
  String recipe = 'hero_protagonico',
  String action = 'none',
  SlotComposition? title,
  String? accentPosition,
  NestedViewComposition? expand,
  String? linkField,
  String? targetRecipe,
  TargetComposition? targetComposition,
  SlotOverrides? slotOverrides,
}) =>
    ViewComposition(
      recipe: recipe,
      action: action,
      slots: {'title': title ?? const SlotComposition(fields: ['nombre'])},
      accentPosition: accentPosition,
      expand: expand,
      linkField: linkField,
      targetRecipe: targetRecipe,
      targetComposition: targetComposition,
      slotOverrides: slotOverrides,
    );

List<FieldDefLike> fieldsBase() => const [
      FieldDefLike(key: 'nombre', type: 'text'),
      FieldDefLike(key: 'edad', type: 'number'),
      FieldDefLike(key: 'avatar', type: 'image'),
      FieldDefLike(key: 'galeria', type: 'array<image>', behavior: 'gallery'),
    ];

ValidateCompositionOptions withFields() =>
    ValidateCompositionOptions(fieldDefs: fieldsBase());

void main() {
  group('happy paths', () {
    test('canónica sin fieldDefs → valid + sin issues', () {
      final r = validateComposition(vc());
      expect(r.valid, true);
      expect(r.issues, isEmpty);
    });

    test('canónica con fieldDefs → sin errores', () {
      final r = validateComposition(vc(), withFields());
      expect(r.issues.where((i) => i.level == 'error'), isEmpty);
    });

    test('accentPosition válido → valid', () {
      expect(validateComposition(vc(accentPosition: 'top')).valid, true);
    });
  });

  group('recipe', () {
    test('recipe inexistente → error', () {
      final r = validateComposition(vc(recipe: 'fake_recipe'));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.path == 'recipe' && i.level == 'error'), true);
    });
  });

  group('action', () {
    test('action inválida → error', () {
      final r = validateComposition(vc(action: 'fake_action'));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.path == 'action' && i.level == 'error'), true);
    });

    test('cada action válida pasa', () {
      for (final a in const [
        'none',
        'navigate_to_detail',
        'modal',
        'expand_inline',
        'external_link'
      ]) {
        final r = validateComposition(vc(action: a));
        expect(r.issues.where((i) => i.path == 'action' && i.level == 'error'),
            isEmpty);
      }
    });
  });

  group('slot fields', () {
    test('field inexistente con fieldDefs → error', () {
      final r = validateComposition(
          vc(title: const SlotComposition(fields: ['fantasma'])), withFields());
      expect(r.valid, false);
      expect(r.issues.any((i) => i.message.contains('fantasma')), true);
    });

    test('field inexistente SIN fieldDefs → no falla', () {
      final r =
          validateComposition(vc(title: const SlotComposition(fields: ['fantasma'])));
      expect(r.issues.where((i) => i.path.startsWith('slots.title.fields')), isEmpty);
    });

    test('field incompatible con accept kind → error', () {
      final r = validateComposition(
          vc(title: const SlotComposition(fields: ['avatar'])), withFields());
      expect(r.valid, false);
      expect(r.issues.any((i) => i.message.contains('no es compatible')), true);
    });

    test('single slot con múltiples fields → warn (no error)', () {
      final r = validateComposition(
          vc(title: const SlotComposition(fields: ['nombre', 'edad'])), withFields());
      expect(r.issues.any((i) => i.level == 'warn' && i.message.contains('single')),
          true);
    });
  });

  group('appearance', () {
    test('shape inválido → error', () {
      final r = validateComposition(vc(
          title: const SlotComposition(
              fields: ['nombre'], appearance: SlotAppearance(shape: 'fake'))));
      expect(r.valid, false);
      expect(
          r.issues.any(
              (i) => i.path == 'slots.title.appearance.shape' && i.level == 'error'),
          true);
    });

    test('aspect inválido → error', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'], appearance: SlotAppearance(aspect: 'fake'))))
              .valid,
          false);
    });

    test('align inválido → error', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'], appearance: SlotAppearance(align: 'fake'))))
              .valid,
          false);
    });

    test('weight inválido → error', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'], appearance: SlotAppearance(weight: 'fake'))))
              .valid,
          false);
    });

    test('size inválido → error', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'], appearance: SlotAppearance(size: 'fake'))))
              .valid,
          false);
    });

    test('truncateChars fuera de rango → error', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'],
                      appearance: SlotAppearance(truncateChars: 1000))))
              .valid,
          false);
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'], appearance: SlotAppearance(truncateChars: 0))))
              .valid,
          false);
    });

    test('truncateChars válido (1-500) → ok', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'], appearance: SlotAppearance(truncateChars: 80))))
              .valid,
          true);
    });

    test('accentPosition (appearance) inválido → error', () {
      expect(
          validateComposition(vc(
                  title: const SlotComposition(
                      fields: ['nombre'],
                      appearance: SlotAppearance(accentPosition: 'fake'))))
              .valid,
          false);
    });

    test('accentPosition top-level inválido → error', () {
      final r = validateComposition(vc(accentPosition: 'fake'));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.path == 'accentPosition' && i.level == 'error'),
          true);
    });
  });

  group('orientation', () {
    test('orientation inválida → error', () {
      expect(
          validateComposition(
                  vc(title: const SlotComposition(fields: ['nombre'], orientation: 'diagonal')))
              .valid,
          false);
    });

    test('orientation horizontal/vertical → ok', () {
      expect(
          validateComposition(
                  vc(title: const SlotComposition(fields: ['nombre'], orientation: 'vertical')))
              .valid,
          true);
    });
  });

  group('expand', () {
    test('expand.recipe inexistente → error', () {
      expect(
          validateComposition(
                  vc(expand: const NestedViewComposition(recipe: 'fake', slots: {})))
              .valid,
          false);
    });

    test('expand.recipe de kind=detail → error', () {
      final r = validateComposition(vc(
          expand: const NestedViewComposition(recipe: 'hero_protagonico', slots: {})));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.message.contains('no es de kind=expand')), true);
    });

    test('expand.recipe de kind=expand → ok', () {
      final r = validateComposition(vc(
          expand: const NestedViewComposition(recipe: 'accordion_simple', slots: {})));
      expect(
          r.issues
              .where((i) => i.path.startsWith('expand.recipe') && i.level == 'error'),
          isEmpty);
    });
  });

  group('targetRecipe', () {
    test('targetRecipe inexistente → error', () {
      expect(validateComposition(vc(targetRecipe: 'fake')).valid, false);
    });

    test('targetRecipe de kind=list → warn (no error)', () {
      final r = validateComposition(vc(targetRecipe: 'compact_avatar'));
      expect(r.issues.any((i) => i.path == 'targetRecipe' && i.level == 'warn'), true);
      expect(r.valid, true);
    });

    test('targetRecipe de kind=detail → ok', () {
      final r = validateComposition(vc(targetRecipe: 'hero_protagonico'));
      expect(r.issues.where((i) => i.path == 'targetRecipe'), isEmpty);
    });
  });

  group('slotOverrides', () {
    test('disabled con slot inexistente → warn', () {
      final r =
          validateComposition(vc(slotOverrides: const SlotOverrides(disabled: ['fantasma'])));
      expect(
          r.issues.any((i) =>
              i.level == 'warn' && i.path.startsWith('slotOverrides.disabled')),
          true);
      expect(r.valid, true);
    });

    test('custom slot shape válido → ok', () {
      final r = validateComposition(vc(
          slotOverrides: const SlotOverrides(custom: [
        CustomSlotDefinition(
            id: 'extra', label: 'Extra', kind: 'single', accepts: ['text-short']),
      ])));
      expect(r.issues.where((i) => i.path.startsWith('slotOverrides.custom')), isEmpty);
    });

    test('custom slot sin label → error', () {
      final r = validateComposition(vc(
          slotOverrides: const SlotOverrides(custom: [
        CustomSlotDefinition(id: 'x', label: '', kind: 'single', accepts: ['text-short']),
      ])));
      expect(r.valid, false);
    });

    test('custom slot con kind inválido → error', () {
      final r = validateComposition(vc(
          slotOverrides: const SlotOverrides(custom: [
        CustomSlotDefinition(id: 'x', label: 'X', kind: 'fake', accepts: ['text-short']),
      ])));
      expect(r.valid, false);
    });
  });

  group('linkField', () {
    test('linkField ausente → ok', () {
      final r = validateComposition(vc(), withFields());
      expect(r.issues.where((i) => i.path == 'linkField'), isEmpty);
    });

    test('linkField string vacío → error', () {
      final r = validateComposition(vc(linkField: '   '));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.path == 'linkField'), true);
    });

    test('linkField referencia field existente → ok', () {
      final r = validateComposition(vc(linkField: 'nombre'), withFields());
      expect(r.issues.where((i) => i.path == 'linkField'), isEmpty);
    });

    test('linkField referencia field inexistente → error', () {
      final r = validateComposition(vc(linkField: 'fantasma'), withFields());
      expect(r.valid, false);
      expect(
          r.issues.any((i) => i.path == 'linkField' && i.message.contains('fantasma')),
          true);
    });

    test('linkField inexistente SIN fieldDefs → no falla', () {
      final r = validateComposition(vc(linkField: 'fantasma'));
      expect(r.issues.where((i) => i.path == 'linkField'), isEmpty);
    });
  });

  group('nestedComposition (depth max=2)', () {
    test('nested válido (depth 2) → ok', () {
      final r = validateComposition(vc(
          title: const SlotComposition(fields: ['nombre'], nestedComposition: NestedViewComposition(
              recipe: 'compact_avatar',
              slots: {'title': SlotComposition(fields: ['otro'])}))));
      expect(
          r.issues.where(
              (i) => i.path.contains('nestedComposition') && i.level == 'error'),
          isEmpty);
    });

    test('nested con recipe vacío → error', () {
      final r = validateComposition(vc(
          title: const SlotComposition(
              fields: ['nombre'],
              nestedComposition: NestedViewComposition(recipe: '', slots: {}))));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.path.endsWith('nestedComposition.recipe')), true);
    });

    test('nested DENTRO de nested → error (depth > 2)', () {
      final r = validateComposition(vc(
          title: const SlotComposition(fields: ['nombre'], nestedComposition: NestedViewComposition(
              recipe: 'compact_avatar',
              slots: {
                'title': SlotComposition(fields: ['x'], nestedComposition: NestedViewComposition(recipe: 'compact_card', slots: {})),
              }))));
      expect(r.valid, false);
      expect(r.issues.any((i) => i.message.contains('profundidad máxima')), true);
    });

    test('nested.slots con field key vacío → error', () {
      final r = validateComposition(vc(
          title: const SlotComposition(fields: ['nombre'], nestedComposition: NestedViewComposition(
              recipe: 'compact_avatar',
              slots: {'title': SlotComposition(fields: ['  '])}))));
      expect(r.valid, false);
    });

    // NOTA: "nested.slots NO es objeto → error" (slots: null en TS) es
    // irrepresentable en Dart: `NestedViewComposition.slots` es
    // `Map<String, SlotComposition>` (no-null) y `fromJson` coacciona null→{}.
    // El parse lo previene antes del validador → garantía más fuerte.
  });

  group('semánticos', () {
    test('valid=true cuando solo hay warnings', () {
      final r = validateComposition(vc(
          slotOverrides: const SlotOverrides(disabled: ['fantasma']),
          targetRecipe: 'compact_avatar'));
      expect(r.issues, isNotEmpty);
      expect(r.issues.every((i) => i.level == 'warn'), true);
      expect(r.valid, true);
    });

    test('valid=false si hay AL MENOS un error', () {
      final r = validateComposition(vc(
          recipe: 'fake', slotOverrides: const SlotOverrides(disabled: ['fantasma'])));
      expect(r.valid, false);
    });

    test('cada issue incluye path + message + level', () {
      final r = validateComposition(vc(recipe: 'fake'));
      for (final i in r.issues) {
        expect(i.path, isNotEmpty);
        expect(i.message, isNotEmpty);
        expect(['error', 'warn'].contains(i.level), true);
      }
    });
  });
}
